### PR TITLE
feat: implement many featuers for the co-brillig rep3 backend

### PR DESCRIPTION
**Pull request by rw0x0**

### CODE REVIEW

File: tests/tests/noir/witness_extension_tests/rep3.rs [hunk 1]
```diff
@@ -9,14 +9,21 @@ use tests::rep3_network::Rep3TestNetwork;
 
 use super::add_rep3_acvm_test;
```
Comment (dkales) [inline tests/tests/noir/witness_extension_tests/rep3.rs:11]: can we enable all that work here?

File: co-noir/co-acvm/src/mpc.rs [hunk 2]
```diff
@@ -25,11 +25,13 @@ pub trait NoirWitnessExtensionProtocol<F: PrimeField> {
 
     type BrilligDriver: BrilligDriver<F>;
 
-    fn init_brillig_driver(&self) -> Self::BrilligDriver;
+    fn init_brillig_driver(&mut self) -> std::io::Result<Self::BrilligDriver>;
 
+    #[expect(clippy::wrong_self_convention)]
     fn from_brillig_result(
```
Comment (dkales) [inline co-noir/co-acvm/src/mpc.rs:31]: this should be called `parse_brillig_result` or smth

File: mpc-core/src/protocols/rep3/arithmetic.rs [hunk 3]
```diff
@@ -469,10 +469,19 @@ pub fn le_public<F: PrimeField, N: Rep3Network>(
     rhs: F,
     io_context: &mut IoContext<N>,
 ) -> IoResult<FieldShare<F>> {
-    let res = detail::unsigned_ge_const_lhs(rhs, lhs, io_context)?;
+    let res = le_public_bit(lhs, rhs, io_context)?;
     conversion::bit_inject(&res, io_context)
 }
 
+/// Same as ge_public but without using bit_inject on the result. Returns 1 if lhs <= rhs and 0 otherwise. Checks if a shared value is less than or equal to a public value. The result is a shared value that has value 1 if the shared value is less than or equal to the public value and 0 otherwise.
```
[the remark this answers is not in the document]
Comment (rw0x0) [reply mpc-core/src/protocols/rep3/arithmetic.rs]: done

File: co-noir/co-brillig/src/mpc/rep3.rs [hunk 4]
```diff
@@ -54,129 +55,1776 @@ impl<F: PrimeField> Default for Rep3BrilligType<F> {
...
+                if let (Public::Field(_), Shared::Field(_)) = (public, shared) {
+                    todo!("Implement IntDiv for public/shared")
+                } else {
+                    eyre::bail!("IntDiv only supported on fields")
+                }
+            }
+            (Rep3BrilligType::Shared(shared), Rep3BrilligType::Public(public)) => {
+                if let (Public::Field(public), Shared::Field(shared)) = (public, shared) {
+                    let divisor: BigUint = public.into();
+                    if divisor.count_ones() == 1 {
```
Comment (dkales) [inline co-noir/co-brillig/src/mpc/rep3.rs:961]: there is a is_power_of_two function that is more descriptive here

**Review by dkales — APPROVED** [review]

**Pull request merged**